### PR TITLE
Adding temperature scaling on Joiner logits:

### PR DESCRIPTION
**Pull request by KarelVesely84**
- T is configurable
- so far best result NCE 0.122 (still not so high) 
  - the BPE scores were rescaled with 0.2 (but then also incorrect words get high confidence, visually reasonable histograms are for 0.5 scale) 
  - BPE->WORD score merging done by min(.) function (tried also prob-product, and also arithmetic, geometric, harmonic mean)

- without temperature scaling (i.e. scale 1.0), the best NCE was 0.032 (here product merging was the best)

Results seem consistent with: https://arxiv.org/abs/2110.15222

Everything tuned on a very small set of 100 sentences with 813 words and 10.2% WER, a Czech model.

I also experimented with blank posteriors mixed into the BPE confidences, but no NCE improvement found, so not pushing that.

Temperature scling added also to the Greedy search confidences.

### CODE REVIEW

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc [hunk 1]
```diff
@@ -129,6 +129,22 @@ void OnlineTransducerModifiedBeamSearchDecoder::Decode(
         model_->RunJoiner(std::move(cur_encoder_out), View(&decoder_out));
 
     float *p_logit = logit.GetTensorMutableData<float>();
+
+    // copy raw logits, apply temperature-scaling (for confidences)
+    int32_t p_logit_items = vocab_size * num_hyps;
+    std::vector<float> logit_with_temperature(p_logit_items);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc:137]: Can we change p_logit in-place?

**Comment by csukuangfj** [comment]
Thanks!

> T hard-coded to 2.0 (is it okay ?)


 Could you make it configurable and give it a default value 1.0 (like what we are doing for blank penalty)?
If it is 1.0, then the loop is skipped.

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc (hunk 1 above)
Comment (KarelVesely84) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc:137]: no, it cannot be done in-place

the idea is to apply temperature only for computation of confidences, 
the decoding continues to use the original values

this is why the logit values are copied to a new buffer

Comment (csukuangfj) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc:137]: Thanks for the explanation.

**Comment by KarelVesely84** [comment]
> Thanks!
> 
> > T hard-coded to 2.0 (is it okay ?)
> 
> Could you make it configurable and give it a default value 1.0 (like what we are doing for blank penalty)? If it is 1.0, then the loop is skipped.

okay, working on it

i am not sure about the default 1.0, for 1.0 the confidences have worse quality than for 2.0,
and, as it is implemented, setting it to 2.0 will not affect the decoding at all, just the confidence
values will be different, to **proposing to set it to 2.0**
(same value appears in the Google article for NN-posterios confidences)

**Comment by csukuangfj** [comment]
1.0 is for backward compatibility.

> as it is implemented, setting it to 2.0 will not affect the decoding at all,

In that case, 2.0 is fine with me.

**Comment by KarelVesely84** [comment]
okay, the T parameter is now configurable

**Comment by KarelVesely84** [comment]
there seems to be some problem with the workload tests, many fail with the 503 error leading to huggingface

**Comment by KarelVesely84** [comment]
i fixed an error in the android build

**Comment by KarelVesely84** [comment]
the tests look OK, seeing only unrelated errors

**Comment by csukuangfj** [comment]
Thank you for your contribution!
